### PR TITLE
[Bugfix] Honor STEP token pooling in DispatchPooler.for_seq_cls

### PR DESCRIPTION
**Pull request by krsish**
## Purpose

`DispatchPooler.for_seq_cls` forces `AllPool()` for `token_classify`, silently ignoring `tok_pooling_type="STEP"`. Remove the override so the existing factory honors the resolved configuration and requests token IDs for STEP. Default/ALL behavior is unchanged.

Added an ALL/STEP regression covering output values, request boundaries, empty matches, and token-ID requirements. No duplicate open PR found; #53555 changes LoRA classifier replacement and #48214 fixes a separate input-buffer race.

## Test Plan

```bash
.venv/bin/python -m pytest tests/model_executor/layers/test_pooler_methods.py -v
pre-commit run --files vllm/model_executor/layers/pooler/special.py tests/model_executor/layers/test_pooler_methods.py
git diff --check
```

## Test Result

- Before: 38 passed; the new STEP case failed. After: **39 passed**.
- Pre-commit on both changed files and `git diff --check`: passed on the GPU validation host.
- GPU validation at `9cd956c7e6cf54efa366b803cafa15ec6c2df827`: `Qwen/Qwen3-ForcedAligner-0.6B-hf` (revision `c07281df297b9905d24a508279258cccf987a064`), `--runner pooling --convert classify --dtype bfloat16 --enforce-eager`, H100, TP=1, `VLLM_BATCH_INVARIANT=1`, activation disabled. Worker inspection confirmed the affected `DispatchPooler` path.

Two public [LibriSpeech audio](https://huggingface.co/datasets/bezzam/audio_samples/resolve/6f4e095fb1a346c4b4f58a5b2d331ebee8ce4d46/librispeech_mr_quilter.wav) inputs (first 2.4 s / full 5.855 s), tested individually and batched:

| Configuration | Before | After |
| --- | --- | --- |
| Default / ALL | `[51,5000]`, `[134,5000]` | unchanged |
| STEP, `step_tag_id=151705` | `[51,5000]`, `[134,5000]` | `[10,5000]`, `[34,5000]` |

STEP matches the corresponding ALL timestamp rows exactly (maximum absolute error `0`, argmax agreement `100%`); default/ALL values are unchanged. Patched `/pooling` returned HTTP 200 and `[10,5000]`, matching offline STEP. This validates output selection, not broader alignment accuracy or throughput.

AI assistance: Codex assisted with implementation, validation, and PR preparation.

### CODE REVIEW

**Comment by noooop** [comment]
cc @haosdent


The current STEP token pooling in vLLM is specifically designed for the Process Reward Model. Is it necessary to support this feature for Qwen/Qwen3-ForcedAligner-0.6B-hf?

**Comment by krsish** [comment]
Thanks for the context. STEP is not required for alignment correctness: ALL followed by client-side filtering works. The motivation is to return only timestamp-position scores from `/pooling`, avoiding transfer of unused token rows.

The existing [online example](https://github.com/vllm-project/vllm/blob/9cd956c7e6cf54efa366b803cafa15ec6c2df827/examples/pooling/token_classify/forced_alignment_online.py#L192-L200) already selects `timestamp_token_id` positions on the client. In our `--convert classify` setup, `step_tag_id=151705` would perform that selection on the server. In the validated example, this reduces the output from `[134,5000]` to `[34,5000]`, with identical values at the selected rows.

This PR preserves default ALL behavior and honors the explicitly configured STEP mode, which this dispatch path currently overrides. We only need the existing tag-based row selection, without `returned_token_ids` or new alignment logic. Would reusing STEP for this opt-in filtering use case fit the intended API?

**Comment by noooop** [comment]
Please update the example to use this new feature.

**Review by Manny7717 — APPROVED** [review]
APPROVE — the hardcoded `AllPool()` override in `DispatchPooler.for_seq_cls`'s `token_classify` entry is removed so `pooler_for_token_classify` honors the resolved `tok_pooling_type` (its `pooling=None` branch already resolves via `get_tok_pooling_method(pooler_config.get_tok_pooling_type())`). Verified locally at head d432a8c4 vs PR base 9cd956c7e.

**Executed verification (VLLM_TARGET_DEVICE=cpu — the file's device-manager paths need an accelerator-less override on this host):**
- Head: full `tests/model_executor/layers/test_pooler_methods.py` = 39/39 pass (ALL and STEP cases both green; without the env override the file is un-runnable here due to the CPU-host autouse cleanup fixture — same class of noise seen on other vLLM PRs).
- Base regression: transplant the new parametrized test onto 9cd956c7e → 38/39, exactly `[STEP]` FAILS: seq0 yields 4 rows (`torch.Size([4, 3]) != [2, 3]`) — all tokens classified instead of the 2 STEP rows (ids 1,3) — and `requires_token_ids` is False. The `[ALL]` case passes on both, so default behavior is byte-identical (resolved ALL → `AllPool()`).
- `ruff check` clean on both changed files.

**Semantics checked:**
- The override was present from the file's original authoring (not a later deliberate workaround), and `for_seq_cls`'s 10+ callers (bert/roberta/modernbert/gpt2/jamba/jina/jina_vl/nemotron_vl, `transformers/pooling.py`, model adapters) route `token_classify` through this factory. Removing the override is a strict alignment with the resolved config: models/deployments whose `tok_pooling_type` default is ALL see zero change; STEP-configured serving now selects step rows AND flips `requires_token_ids` so the engine actually requests token ids (the missing half of the bug — before, a STEP-configured token-classify deployment silently classified every token without ever asking for ids).
- `get_tok_pooling_method` accepts only ALL/STEP (config typed as `TokenPoolingType` literal), and `PoolerConfig` already treats STEP as a first-class combination with seq_cls elsewhere (chunked-prefill gating in `config/model.py`), so honoring it here is consistent with existing contracts.
- The test's row expectations (`[[1,3],[5],[]]` for STEP vs all rows for ALL) exercise the observable output shape through the real `DispatchPooler.forward`, not a mock of selection.

No issues found. The single non-blocking observation: the PR body says "Default/ALL behavior is unchanged" — confirmed by the parametrized ALL case passing on both base and head.

**Review by sylvesterkaczmarek — COMMENTED** [review]
Removing the forced `AllPool()` is the right ownership fix: `pooler_for_token_classify` should resolve the configured token-pooling policy itself. The parameterized ALL/STEP test is useful because it verifies both row selection and the `requires_token_ids` contract, not just the output count.

**Comment by krsish** [comment]
Updated the existing online forced-alignment example in e3634b8:

- it now uses `Qwen/Qwen3-ForcedAligner-0.6B-hf` with `--convert classify` and STEP pooling;
- `151705` is documented as this checkpoint's `timestamp_token_id` from `config.json`;
- the client consumes the returned timestamp rows directly instead of filtering ALL-token logits.

@noooop, could you please take another look?

**Review by noooop — APPROVED** [review]
